wgsl/index.bs

```diff
@@ -2751,29 +2929,116 @@ ISSUE: Can we query upper bounds on workgroup size dimensions?  Is it independen
+
+A <dfn noexport>pipeline input</dfn> is data provided to the shader stage from upstream in the pipeline.
+A pipeline input is denoted by a module-scope variable in the `in` [=storage class=].
+The store type must be [=IO-shareable=].
+
+A <dfn noexport>pipeline output</dfn> is data the shader provides for further processing downstream in the pipeline.
+A pipeline output is denoted by a module-scope variable in the `out` [=storage class=].
+The store type must be [=IO-shareable=].
+
+Each pipeline input or output is one of:
+
+* A built-in variable. See [[#builtin-inputs-outputs]].
+* A user data attribute. See [[#user-data-attributes]].
+
+#### Built-in inputs and outputs #### {#builtin-inputs-outputs}
+
+A <dfn noexport>built-in input variable</dfn> provides access to system-generated control information.
+The set of built-in inputs are listed in [[#builtin-variables]].
+
+To declare a variable for accessing a particular input built-in *X*:
+
+* Declare a module-scope variable in the `in` [=storage class=],
+    where the [=store type=] is the listed store type for *X*.
+* Apply a `builtin(`*X*`)` attribute to the variable.
+* The variable must not have an initializer. The system provides the value.
+
+A <dfn noexport>built-in output variable</dfn> is used by the shader to convey
+control information to later processing steps in the pipeline.
+The set of built-in outputs are listed in [[#builtin-variables]].
+
+To declare a variable for accessing a particular output built-in *Y*:
+
+* Declare a module-scope variable in the `out` [=storage class=],
+    where the [=store type=] is the listed store type for *Y*.
+* Apply a `builtin(`*Y*`)` attribute to the variable.
+* The variable may have an initializer, or not, as described in [[#variables]].
+
+<div class='example' heading="Declaring built-in variables">
+  <xmp>
+    # vertex shader output builtin
+    [[builtin(position)]] var<out> my_position : vec4<f32>;
+
+    # fragment shader input builtin
+    [[builtin(frag_coord)]] var<in> coord : vec4<f32>;
+
+    # compute shader builtin
+    [[builtin(global_invocation_id)]] var<in> global_id : vec3<u32>;
+  </xmp>
+</div>
+
+The `builtin` attribute must not be applied to a variable in a storage class other than `in` or `out`.
 
-### Built-in variables TODO ### {#builtin-var-interface}
+An input built-in must only be applied to a variable in the `in` storage class.
 
-### Pipeline Input and Output Interface TODO ### {#pipeline-inputs-outputs}
+An output built-in must only be applied to a variable in the `out` storage class.
 
-TODO(dneto): The following sentence was moved from elsewhere. Expand this.
+A variable must not have more than one `builtin` attribute.
 
-The input and output parameters to the entry point are determined by which
-global variables are used in the function and any called functions.
+Each built-in variable has an associated shader stage, as described in [[#builtin-variables]].
```

Comment:
I'm hesitating to raise this note... the statement makes it clear that `in`/`out` storage classes for builtins are redundant. Maybe leave an `Issue` in place for this?

Reply:
In Vulkan the builtin variables count against the I/O location count. 
I'll add an Issue.

Reply:
Resolving for this PR, as I've added the Issue to the text.